### PR TITLE
Convert basic DeepSpeech API to v0.4.0

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

**Comment by lissyx** [comment]
```
$ LD_LIBRARY_PATH=/home/alex/tmp/deepspeech/tc-linux/ ~/go/bin/deepspeech ~/tmp/deepspeech/output_graph.fr.pb ~/tmp/deepspeech/test-alex.fr.wav ~/codaz/Mozilla/DeepSpeech/deepspeech-kdavis/data/alphabet.txt 
TensorFlow: v1.12.0-10-ge232881
DeepSpeech: v0.4.0-alpha.3-4-ge549f89
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2018-12-20 15:52:33.605582: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
INFO[0003] Text: monjur sette bois                      
```

**Comment by lissyx** [comment]
@asticode I've opened that to get some early feedback, this does not yet include the Streaming API.

I don't know Go very well, it would be nice if we could change the CLI args handling and have some positional arguments as we did everywhere now :).

Also, does go offer a way to bundle the `.so` directly with the binary?

**Comment by asticode** [comment]
Thanks for this!

3 things:

- right now, https://github.com/mozilla/DeepSpeech/releases/download/v0.4.0/audio-0.4.0.tar.gz returns a `404`, is that normal?

- when I run `LD_LIBRARY_PATH="/tmp/ds" deepspeech models/output_graph.pb audio/2830-3980-0043.wav models/alphabet.txt models/lm.binary models/trie` (with my models + audio being from `v0.3.0` though), I get the following:

```
TensorFlow: v1.12.0-10-ge232881c5a
DeepSpeech: unknown
Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage.
2018-12-20 16:38:22.197011: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Error: Trie file version mismatch (2 instead of expected 3). Update your trie file.
INFO[0010] Text:  a a a a a a a a a a a a a a a  a a a a a
```

Any idea why ? (except because models and audios are from `v0.3.0`)

- what do you mean by `positional arguments`? can you point to the corresponding C files/lines?

Cheers

**Comment by lissyx** [comment]
>     * right now, https://github.com/mozilla/DeepSpeech/releases/download/v0.4.0/audio-0.4.0.tar.gz returns a `404`, is that normal?

Yes


> Any idea why ? (except because models and audios are from `v0.3.0`)

Normal, 0.3.0 model needs to be re-exported with a softmax layer for the new decoder, and lots of people are getting tricked by that right now, that's also why we are going to release 0.4.0 asap

> what do you mean by `positional arguments`? can you point to the corresponding C files/lines?

Indeed, I have the patch ready in fact

**Comment by asticode** [comment]
Regarding flags, I'd rather not use a dependency to parse them since the standard library does it pretty well itself:

```go
var model = flag.String("m", "model", "Path to the model (protocol buffer binary file)")
// Declare other flags outside of the main function

func main() {
  flag.Parse()
}
```

And the rest can stay the same.

If you have a URL pointing to valid 0.4.0 models so that I can test, I'll be waiting for the flags correction to merge the PR.

**Comment by lissyx** [comment]
> If you have a URL pointing to valid 0.4.0 models so that I can test, I'll be waiting for the flags correction to merge the PR.

I'm working on adding the streaming API as well. For the 0.4.0 model, we don't have anything yet online, but I can share you one french one kinda working.

**Comment by lissyx** [comment]
@asticode Looking at the doc https://golang.org/pkg/flag/ I cannot find how to set required / non required flags, if you have a quick pointer to that :)

**Comment by asticode** [comment]
You can't add required/non required flags but I don't think this is a big issue in that case.

Got it about the model, I guess I'll need to wait for one to be available online before merging the PR. What do you think?

**Comment by lissyx** [comment]
> Got it about the model, I guess I'll need to wait for one to be available online before merging the PR. What do you think?

Are you worrying about stale links ?

> You can't add required/non required flags but I don't think this is a big issue in that case.

That's unfortunate, that's why `argparse` looked better in my eyes, since can be optional

**Comment by lissyx** [comment]
@asticode You should be able to pull a working `.pb` or `.pbmm` from https://github.com/reuben/DeepSpeech/releases/tag/v0.2.0-prod-ctcdecode

**Comment by lissyx** [comment]
I'm also reverting to `flag` package

**Comment by lissyx** [comment]
@asticode Now should be better, with `flag` package instead of `argparse`

**Comment by lissyx** [comment]
@asticode I'm being blocked on how to expose `StreamingState` to the user, do you have any suggestion for that?

**Comment by asticode** [comment]
I'm not familiar with `StreamingState`. Is it a method or an attribute? Could you link to its C declaration? What problem are you facing?

**Comment by lissyx** [comment]
> I'm not familiar with `StreamingState`. Is it a method or an attribute? Could you link to its C declaration? What problem are you facing?

Just a `struct`: https://github.com/mozilla/DeepSpeech/blob/v0.4.0-alpha.3/native_client/deepspeech.h#L20
https://github.com/mozilla/DeepSpeech/blob/v0.4.0-alpha.3/native_client/deepspeech.h#L112
https://github.com/mozilla/DeepSpeech/blob/v0.4.0-alpha.3/native_client/deepspeech.h#L139

So I just cannot find how in Go I should expose that ; right now, it's within the `ModelWrapper` class, but that's not really the appropriate way.

**Comment by asticode** [comment]
You should be able to declare it with:

```go
type StreamingState C.struct_StreamingState
```

Check out [this link](https://github.com/asticode/goav/blob/master/avcodec/avcodec.go#L26) for an example.

Although this is pure C whereas Deepspeech is C++ so there may be some more tweaking to do.

**Comment by lissyx** [comment]
> You should be able to declare it with:
> 
> ```go
> type StreamingState C.struct_StreamingState
> ```
> 
> Check out [this link](https://github.com/asticode/goav/blob/master/avcodec/avcodec.go#L26) for an example.
> 
> Although this is pure C whereas Deepspeech is C++ so there may be some more tweaking to do.

Instead I exposed a `StreamWrapper`, but I had to make the `ModelState` a public member. Not sure how this can be an issue wrt Go best practices.

**Comment by asticode** [comment]
Can't you keep the attribute private and expose a public getter instead?

**Comment by lissyx** [comment]
> Can't you keep the attribute private and expose a public getter instead?

Why not ...

**Comment by lissyx** [comment]
@asticode Should be done, if you can give that a quick look, then I'll ensure the documentation is uptodate as well.

Thanks for your reactivity!

File: README.md
Comment (asticode) [inline README.md]: Is it safe to use the `v0.4.0` release instead of `v0.4.0-alpha.3`?

File: README.md
Comment (asticode) [inline README.md]: Is it safe to use the `v0.4.0` release instead of `v0.4.0-alpha.3`?

File: deepspeech_wrap.h
Comment (asticode) [inline deepspeech_wrap.h:1]: Is there a reason why you renamed the file to `deepspeech_wrap` ?

**Review by asticode — CHANGES_REQUESTED** [review]
This PR seems awesome! Just a few minor questions though.

File: deepspeech_wrap.h
Comment (lissyx) [reply deepspeech_wrap.h:1]: it was colliding with our `deepspeech.h`, renaming ensures people cannot hit that and can build with just adjusting `CFLAGS` and `LDFLAGS`

File: README.md
Comment (lissyx) [reply README.md]: I've used v0.4.0-alpha.3 there to ensure you could fetch and test, v0.4.0 is not yet a tag. Once it is, it should be used instead of v0.4.0-alpha.3. It should be a matter of days now.

Comment (asticode) [reply README.md]: OK, I'd rather you update the tag now in the PR so that we don't forget if you don't mind. I'm afraid we'll forget to change it afterwards...

Comment (lissyx) [reply README.md]: @asticode Should be `v0.4.0` now :)

**Comment by asticode** [comment]
Perfect, just merged the PR!

Thanks again!

**Pull request closed without merging**
